sos/plugins/vdsm.py

```diff
@@ -0,0 +1,151 @@
+# there is a stuck lvm command.
+# locking_type=0
+#
+# use_lvmetad is set to 0 in order not to show cached, old lvm metadata.
+# use_lvmetad=0
+#
+# preferred_names and filter config values are set to capture Vdsm devices.
+# preferred_names=[ '^/dev/mapper/' ]
+# filter=[ 'a|^/dev/mapper/.*|', 'r|.*|' ]
+LVM_CONFIG = """
+global {
+    locking_type=0
+    use_lvmetad=0
+}
+devices {
+    preferred_names=["^/dev/mapper/"]
+    ignore_suspended_devices=1
+    write_cache_state=0
+    disable_after_error_count=3
+    filter=["a|^/dev/mapper/.*|", "r|.*|"]
+}
+"""
+LVM_CONFIG = re.sub(r"\s+", " ", LVM_CONFIG).strip()
+
+
+class Vdsm(Plugin, RedHatPlugin):
+    """vdsm Plugin"""
+
+    packages = (
+        'vdsm',
+        'vdsm-client',
+    )
+
+    plugin_name = 'vdsm'
+
+    def setup(self):
+        lvm_system_dir = os.environ["LVM_SYSTEM_DIR"]
+        os.environ["LVM_SYSTEM_DIR"] = "/var/run/vdsm/lvm"
+        self.add_forbidden_path('/etc/pki/vdsm/keys/*')
+        self.add_forbidden_path('/etc/pki/vdsm/libvirt-spice/*-key.*')
+        self.add_forbidden_path('/etc/pki/libvirt/private/*')
+
+        self.add_cmd_output('service vdsmd status')
+
+        self.add_copy_spec([
+            '/tmp/vds_installer*',
+            '/tmp/vds_bootstrap*',
+            '/etc/vdsm/*'
+        ])
+
+        limit = self.get_option('log_size')
+
+        self.add_copy_spec('/var/log/vdsm/*', sizelimit=limit)
+
+        self._add_vdsm_forbidden_paths()
+        self.add_copy_spec([
+            '/var/run/vdsm/*',
+            '/usr/libexec/vdsm/hooks',
+            '/var/lib/vdsm'
+        ])
+
+        qemu_pid = self.get_process_pid('qemu-kvm')
+        if qemu_pid:
+            self.add_copy_spec([
+                "/proc/%s/cmdline" % qemu_pid,
+                "/proc/%s/status" % qemu_pid,
+                "/proc/%s/mountstats" % qemu_pid
+            ])
+        self.add_cmd_output([
+            "ls -ldZ /etc/vdsm",
+            "su vdsm -s sh -c 'tree -l /rhev/data-center'",
+            "su vdsm -s sh -c 'ls -lR /rhev/data-center'"
+        ])
+        self.add_cmd_output([
+            "lvm vgs -v -o +tags --config \'%s\'" % LVM_CONFIG,
+            "lvm lvs -v -o +tags --config \'%s\'" % LVM_CONFIG,
+            "lvm pvs -v -o +all --config \'%s\'" % LVM_CONFIG
+        ])
+
+        self.add_cmd_output([
+            'vdsm-client Host getCapabilities',
+            'vdsm-client Host getStats',
+            'vdsm-client Host getAllVmStats',
+            'vdsm-client Host getVMFullList',
+            'vdsm-client Host getDeviceList',
+            'vdsm-client Host hostdevListByCaps',
+            'vdsm-client Host getAllTasksInfo',
+            'vdsm-client Host getAllTasksStatuses'
+        ])
+
+        try:
+            res = self.call_ext_prog(
+                'vdsm-client Host getConnectedStoragePools'
+            )
+            if res['status'] == 0:
+                pools = ast.literal_eval(res['output'])
```

Comment:
This is wrong. The output of vdsm-client commands is json, not python dict representation,
so it should be parsed with json.loads().

Reply:
Thanks. I thought some of these looked suspicious (and also missing exception handling).